File: mns-reactor/src/main/java/cz/fb/manaus/reactor/filter/UnprofitableCategoriesRefresher.java
package cz.fb.manaus.reactor.filter;

import cz.fb.manaus.core.maintanance.ConfigUpdate;
import cz.fb.manaus.core.maintanance.PeriodicMaintenanceTask;
import cz.fb.manaus.spring.ManausProfiles;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Component;

import java.time.Duration;
import java.util.LinkedList;
import java.util.List;

@Component
@Profile(ManausProfiles.DB)
public class UnprofitableCategoriesRefresher implements PeriodicMaintenanceTask {

    public static final String REFRESH_PERIOD_EL = "#{systemEnvironment['MNS_UNPROFITABLE_REFRESH_PERIOD_HRS'] ?: 8}";
    private final int refreshPeriodHours;

    @Autowired
    public UnprofitableCategoriesRefresher(@Value(REFRESH_PERIOD_EL) int refreshPeriodHours) {
        this.refreshPeriodHours = refreshPeriodHours;
    }

    @Autowired(required = false)
    private List<AbstractUnprofitableCategoriesRegistry> unprofitableCategoriesRegistries = new LinkedList<>();

    @Override
    public String getName() {
        return "unprofitableCategoriesRefresh";
    }

    @Override
    public Duration getPausePeriod() {
        return Duration.ofHours(8);
    }

    @Override
    public ConfigUpdate execute() {
        var configUpdate = ConfigUpdate.empty(Duration.ofDays(1));
        for (var registry : unprofitableCategoriesRegistries) {
            registry.updateBlacklists(configUpdate);
        }
        return configUpdate;
    }
}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/filter/UnprofitableCategoriesRefresher.kt
package cz.fb.manaus.reactor.filter

import cz.fb.manaus.core.maintanance.PeriodicTask
import cz.fb.manaus.core.repository.BlacklistedCategoryRepository
import cz.fb.manaus.spring.ManausProfiles
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Component
import java.time.Duration
import java.util.logging.Logger

@Component
@Profile(ManausProfiles.DB)
class UnprofitableCategoriesRefresher(
        private val blacklistedCategoryRepository: BlacklistedCategoryRepository,
        private val registries: List<AbstractUnprofitableCategoriesRegistry>) : PeriodicTask {

    private val log = Logger.getLogger(UnprofitableCategoriesRefresher::class.simpleName)

    override val name: String = "unprofitableCategoriesRefresh"

    override val pausePeriod: Duration = Duration.ofHours(8)

    override fun execute() {
        val old = blacklistedCategoryRepository.list()
        val current = registries.flatMap { it.getBlacklist() }
        val currentNames = current.map { it.name }.toSet()
        old.filter { it.name !in currentNames }
                .onEach { log.info { "deleting blacklisted category '$it'" } }
                .forEach { blacklistedCategoryRepository.delete(it.name) }
        current.onEach { log.info { "saving blacklisted category '$it'" } }
                .forEach { blacklistedCategoryRepository.saveOrUpdate(it) }
    }

}
